README.md
# @jvlk/fp-ts-remote-data

A library to make it easy to work with remote data. Compatible with the `fp-ts` ecosystem, but it's not required.

RemoteData represents a value that can be in one of four states:

- Loading
- Error
- Empty
- Success

[Read the full documentation](https://jderochervlk.github.io/fp-ts-remote-data/).

## Example in React

```tsx
import * as D from '@jvlk/fp-ts-remote-data'
import { pipe } from 'fp-ts/function'

import useBlogRequest from './useBlogRequest'
import Loading from './Loading'
import Error from './Error'
import Empty from './Empty'
import Blog from './Blog'

function Component() {
  const blog = useBlogRequest()

  return pipe(
    blog,
    D.fold(
        () => <Loading />),
        err => <Error error={err}>,
        () => <Empty/>,
        entries => <Blog entries={entries}>
  )
}
```

## Installation

Using NPM:

```
npm i @jvlk/fp-ts-remote-data
```

Using Yarn:

```
yarn add @jvlk/fp-ts-remote-data
```

## Thanks

The start of this repo began as a clone of https://github.com/sylcastaing/fp-ts-remote-data. I have modified it and diverged from it. That library is much more "sound" from a `fp-ts` types perspective, but I wanted to create something that was focused more on developer usability. I've removed many of the functions in favor of simplifying things. I have also added the 4th state of `empty` to avoid having to use a value of `Option` if the results don't contain any data.

The original concept of the RemoteData type comes from [How Elm Slays a UI Antipattern](http://blog.jenkster.com/2016/06/how-elm-slays-a-ui-antipattern.html).


CHANGELOG.md
# [1.1.0](https://github.com/jderochervlk/fp-ts-remote-data/compare/v1.0.1...v1.1.0) (2023-08-02)


### Features

* remove extra fold function ([77444ce](https://github.com/jderochervlk/fp-ts-remote-data/commit/77444ceebc937d2b1663ca1cb6adafc15afdac3e))

## [1.0.1](https://github.com/jderochervlk/fp-ts-remote-data/compare/v1.0.0...v1.0.1) (2022-11-08)


### Bug Fixes

* remove lib from fp-ts imports ([0fb7c27](https://github.com/jderochervlk/fp-ts-remote-data/commit/0fb7c27efa35af0c73c0ab590d646f5fd509f4d2))

# 1.0.0 (2022-11-04)


### Bug Fixes

* alpha.2 ([86b4720](https://github.com/jderochervlk/fp-ts-remote-data/commit/86b4720c38c0832e067588b2881aa614c3a4eaad))


### Features

* basic docs are complete ([3acf776](https://github.com/jderochervlk/fp-ts-remote-data/commit/3acf77642ca98b29df7a4e00d720ca3d705bddf8))
* v1 alpha1 ([b26ad8f](https://github.com/jderochervlk/fp-ts-remote-data/commit/b26ad8f3e4044bafaf61d853f9d48bfe104422e9))


src/test.ts
/**
 * @internal
 * @since 1.0.0
 */

import * as RD from '.'
import { pipe } from 'fp-ts/function'

test('loading', () => {
  expect(RD.isLoading(RD.loading)).toBeTruthy()
})

test('failure', () => {
  expect(RD.isFailure(RD.failure(Error('Oh no')))).toBeTruthy()
})

test('empty', () => {
  expect(RD.isEmpty(RD.empty)).toBeTruthy()
})

test('success', () => {
  expect(RD.success(RD.success(42))).toBeTruthy()
})

test('map', () => {
  expect(
    pipe(
      RD.success(10),
      RD.map((n) => n + 1),
      RD.map((n) => `${n}`),
      RD.map((str) => str.toLowerCase())
    )
  ).toEqual({ _tag: 'Success', success: '11' })

  expect(
    pipe(
      RD.empty,
      RD.map((n) => n + 1)
    )
  ).toEqual({ _tag: 'Empty' })

  expect(
    pipe(
      RD.failure('oh no'),
      RD.map((n) => n + 1)
    )
  ).toEqual({ _tag: 'Failure', failure: 'oh no' })

  expect(
    pipe(
      RD.loading,
      RD.map((n) => n + 1)
    )
  ).toEqual({ _tag: 'Loading' })
})

test('chain', () => {
  expect(
    pipe(
      RD.success(10),
      RD.chain((_n) => RD.loading)
    )
  ).toEqual({ _tag: 'Loading' })

  expect(
    pipe(
      RD.success(10),
      RD.chain((_n) => RD.empty)
    )
  ).toEqual({ _tag: 'Empty' })

  expect(
    pipe(
      RD.loading,
      RD.chain((n) => RD.empty)
    )
  ).toEqual({ _tag: 'Loading' })

  expect(
    pipe(
      RD.empty,
      RD.chain((n) => RD.loading)
    )
  ).toEqual({ _tag: 'Empty' })

  expect(
    pipe(
      RD.failure('oh no'),
      RD.chain<string>((_n) => RD.loading)
    )
  ).toEqual({ _tag: 'Failure', failure: 'oh no' })

  expect(
    pipe(
      RD.failure('oh no'),
      RD.chainW((_n) => RD.loading)
    )
  ).toEqual({ _tag: 'Failure', failure: 'oh no' })
})

test('reduce', () => {
  expect(
    pipe(
      RD.success(42),
      RD.reduce(10, (acc, item) => acc + item)
    )
  ).toEqual(52)

  expect(
    pipe(
      RD.empty,
      RD.reduce(10, (acc, item) => acc + item)
    )
  ).toEqual(10)

  expect(
    pipe(
      RD.success('foo'),
      RD.reduce('bar', (a, b) => a + b)
    )
  ).toEqual('barfoo')
})

test('reduceRight', () => {
  expect(
    pipe(
      RD.success('foo'),
      RD.reduceRight('bar', (a, b) => a + b)
    )
  ).toEqual('foobar')

  expect(
    pipe(
      RD.empty,
      RD.reduceRight('bar', (a, b) => a + b)
    )
  ).toEqual('bar')
})

test('fold', () => {
  expect(
    pipe(
      RD.loading,
      RD.fold(
        () => 'loading',
        () => 'error',
        () => 'empty',
        () => 'success'
      )
    )
  ).toEqual('loading')

  expect(
    pipe(
      RD.failure(42),
      RD.fold(
        () => 'loading',
        () => 'error',
        () => 'empty',
        () => 'success'
      )
    )
  ).toEqual('error')

  expect(
    pipe(
      RD.empty,
      RD.fold(
        () => 'loading',
        () => 'error',
        () => 'empty',
        () => 'success'
      )
    )
  ).toEqual('empty')

  expect(
    pipe(
      RD.success(42),
      RD.fold(
        () => 'loading',
        () => 'error',
        () => 'empty',
        () => 'success'
      )
    )
  ).toEqual('success')
})

test('foldW can return different types', () => {
  expect(
    pipe(
      RD.loading,
      RD.foldW(
        () => 10,
        console.log,
        () => 'empty',
        () => ['success']
      )
    )
  ).toEqual(10)
})

test('of', () => {
  expect(RD.of({ loading: true })).toEqual(RD.loading)
  expect(RD.of({ loading: true, failure: 'no' })).toEqual(RD.loading)
  expect(RD.of({ loading: true, value: 'no' })).toEqual(RD.loading)
  expect(RD.of({ loading: true, value: 'testing', failure: 'no' })).toEqual(
    RD.loading
  )
  expect(RD.of({ loading: true, value: undefined, failure: 'no' })).toEqual(
    RD.loading
  )

  expect(RD.of({ failure: 'oh no', loading: false })).toEqual(
    RD.failure('oh no')
  )
  expect(RD.of({ loading: false, failure: 'oh no' })).toEqual(
    RD.failure('oh no')
  )
  expect(RD.of({ failure: 'oh no', value: 'oh no', loading: false })).toEqual(
    RD.failure('oh no')
  )
  expect(RD.of({ loading: false, value: 'testing', failure: 'oh no' })).toEqual(
    RD.failure('oh no')
  )
  expect(RD.of({ loading: false, value: undefined, failure: 'oh no' })).toEqual(
    RD.failure('oh no')
  )

  expect(RD.of({ loading: false, value: undefined })).toEqual(RD.empty)

  expect(RD.of({ loading: false, value: 42 })).toEqual(RD.success(42))
})


src/refinements.ts
/**
 * @internal
 * @since 1.0.0
 */

import { Empty, Failure, Loading, RemoteData, Success } from './model'

/**
 * Returns `true` if the RemoteData is an instance of `Loading`, `false` otherwise.
 *
 * @example
 * import * as RD from '@jvlk/fp-ts-remote-data'
 *
 * RD.isLoading(RD.success(1)) // false
 * RD.isLoading(RD.failure(1)) // false
 * RD.isLoading(RD.empty) // false
 * RD.isLoading(RD.loading) // true
 *
 * @category guards
 * @since 1.0.0
 */
export const isLoading = <E, A>(t: RemoteData<E, A>): t is Loading =>
  t._tag === 'Loading'

/**
 * Returns `true` if the RemoteData is an instance of `Failure`, `false` otherwise.
 *
 * @example
 * import * as RD from '@jvlk/fp-ts-remote-data'
 *
 * RD.isFailure(RD.success(1)) // false
 * RD.isFailure(RD.failure(1)) // true
 * RD.isFailure(RD.empty) // false
 * RD.isFailure(RD.loading) // false
 *
 * @category guards
 * @since 1.0.0
 */
export const isFailure = <E, A>(t: RemoteData<E, A>): t is Failure<E> =>
  t._tag === 'Failure'

/**
 * Returns `true` if the RemoteData is an instance of `Empty`, `false` otherwise.
 *
 * @example
 * import * as RD from '@jvlk/fp-ts-remote-data'
 *
 * RD.isEmpty(RD.success(1)) // false
 * RD.isEmpty(RD.failure(1)) // false
 * RD.isEmpty(RD.empty) // true
 * RD.isEmpty(RD.loading) // false
 *
 * @category guards
 * @since 1.0.0
 */
export const isEmpty = <E, A>(t: RemoteData<E, A>): t is Empty =>
  t._tag === 'Empty'

/**
 * Returns `true` if the RemoteData is an instance of `Success`, `false` otherwise.
 *
 * @example
 * import * as RD from '@jvlk/fp-ts-remote-data'
 *
 * RD.isSuccess(RD.success(1)) // true
 * RD.isSuccess(RD.failure(1)) // false
 * RD.isSuccess(RD.empty) // false
 * RD.isSuccess(RD.loading) // false
 *
 * @category guards
 * @since 1.0.0
 */
export const isSuccess = <E, A>(t: RemoteData<E, A>): t is Success<A> =>
  t._tag === 'Success'


src/constructors.ts
/**
 * Functions to create a new RemoteData.
 *
 * @since 1.0.0
 */
import type { RemoteData } from './model'

/**
 * Constructs a new `RemoteData` with a loading state.
 *
 * @example
 * import * as RD from '@jvlk/fp-ts-remote-data'
 *
 * const loadingValue = RD.loading
 *
 * @category constructors
 * @since 1.0.0
 */
export const loading: RemoteData<never, never> = { _tag: 'Loading' }

/**
 * Constructs a new `RemoteData` holding a `Success` value.
 *
 * @example
 * import * as RD from '@jvlk/fp-ts-remote-data'
 *
 * const successValue = RD.success(42)
 *
 * @category constructors
 * @since 1.0.0
 */
export const success = <E = never, A = never>(a: A): RemoteData<E, A> => ({
  _tag: 'Success',
  success: a,
})

/**
 * Constructs a new `RemoteData` holding an `Failure` value.
 *
 * @example
 * import * as RD from '@jvlk/fp-ts-remote-data'
 *
 * const failure = RD.failure(42)
 *
 * @category constructors
 * @since 1.0.0
 */
export const failure = <E = never, A = never>(e: E): RemoteData<E, A> => ({
  _tag: 'Failure',
  failure: e,
})

/**
 * Constructs a new `RemoteData` with an empty state.
 *
 * @example
 * import * as RD from '@jvlk/fp-ts-remote-data'
 *
 * const emptyValue = RD.empty
 *
 * @category constructors
 * @since 1.0.0
 */
export const empty: RemoteData<never, never> = { _tag: 'Empty' }

type remoteDataOf<E, A> = {
  loading: boolean
  failure?: E
  value?: A | null | undefined
}

/**
 * Constructs a new `RemoteData` from an object
 *
 * @category constructors
 * @since 1.0.0
 *
 * @example
 * import * as RD from '@jvlk/fp-ts-remote-data'
 *
 * const one = RD.of({ loading: true }) // => loading
 * const two = RD.of({ loading: false, failure: Error("oh no") }) // => failure
 * const three = RD.of({ loading: false, value: undefined }) // => empty
 * const four = RD.of({ loading: false, value: 42 }) // => success
 */
export const of = <E, A>(obj: remoteDataOf<E, A>): RemoteData<E, A> => {
  return obj.loading
    ? loading
    : obj.failure
    ? failure(obj.failure)
    : obj.value
    ? success(obj.value)
    : empty
}


src/instances.ts
/**
 * @since 1.0.0
 */

import { success } from './constructors'
import { RemoteData } from './model'
import { isEmpty, isFailure, isLoading, isSuccess } from './refinements'

/**
 * @example
 * import * as D from '@jvlk/fp-ts-remote-data'
 * import { pipe } from 'fp-ts/function'
 *
 * pipe(
 *  D.success(42),
 *  D.map(n => n + 10) // => success(52)
 * )
 *
 * pipe(
 *  D.empty,
 *  D.map(n => n + 10) // => empty
 * )
 *
 * @category instance operations
 * @since 1.0.0
 */
export const map: <A, B>(
  f: (a: A) => B
) => <E>(fa: RemoteData<E, A>) => RemoteData<E, B> = (f) => (fa) =>
  isSuccess(fa) ? success(f(fa.success)) : fa

/**
 * Less strict version of [`chain`](#chain) that allows you to **W**iden the failure type.
 *
 * @example
 * import * as D from '@jvlk/fp-ts-remote-data'
 * import { pipe } from 'fp-ts/function'
 *
 * pipe(
 *  D.failure<string, string>('error'), // RemoteData<string, string>
 *  D.chainW(() => D.failure(Error('number is too small'))) // => RemoteData<Error | string, string>
 * )
 *
 * @category instance operations
 * @since 1.0.0
 */
export const chainW =
  <D, A, B>(f: (a: A) => RemoteData<D, B>) =>
  <E>(ma: RemoteData<E, A>): RemoteData<D | E, B> =>
    isSuccess(ma) ? f(ma.success) : ma

/**
 * @example
 * import * as D from '@jvlk/fp-ts-remote-data'
 * import { pipe } from 'fp-ts/function'
 *
 * pipe(
 *  D.success(42),
 *  D.chain(n => n > 10 ? D.success(n) : D.failure(Error('number is too small')))
 * )
 *
 * @category instance operations
 * @since 1.0.0
 */
export const chain: <E, A = E, B = A>(
  f: (a: A) => RemoteData<E, B>
) => (ma: RemoteData<E, A>) => RemoteData<E, B> = chainW

/**
 * @category instance operations
 * @since 1.0.0
 */
export const reduce: <A, B>(
  b: B,
  f: (a: B, b: A) => B
) => <E>(fa: RemoteData<E, A>) => B = (b, f) => (fa) =>
  isSuccess(fa) ? f(b, fa.success) : b

/**
 * @category instance operations
 * @since 1.0.0
 */
export const reduceRight: <A, B>(
  b: B,
  f: (a: A, b: B) => B
) => <E>(fa: RemoteData<E, A>) => B = (b, f) => (fa) =>
  isSuccess(fa) ? f(fa.success, b) : b

/**
 * Fold and return different types in the response.
 *
 * @example
 * import * as D from '@jvlk/fp-ts-remote-data'
 * import { pipe } from 'fp-ts/function'
 *
 * const resultOne = pipe(
 *      D.success(42),
 *      D.foldW(
 *          () => "loading",
 *          e => Error(e),
 *          () => null,
 *          t => [t]
 *      )
 * )
 *
 * @category instance operations
 * @since 1.0.0
 */
export const foldW =
  <E, A, U, T, V, Z>(
    loading: () => U,
    failure: (e: E) => T,
    empty: () => V,
    success: (a: A) => Z
  ) =>
  (d: RemoteData<E, A>) => {
    return isLoading(d)
      ? loading()
      : isFailure(d)
      ? failure(d.failure)
      : isEmpty(d)
      ? empty()
      : success(d.success)
  }

/**
 * Fold and return the same type in the response.
 *
 * @example
 * import * as D from '@jvlk/fp-ts-remote-data'
 * import { pipe } from 'fp-ts/function'
 *
 * const resultOne = pipe(
 *      D.success(42),
 *      D.fold(
 *          () => "loading",
 *          e => `${e}`,
 *          () => "empty",
 *          t => `value is ${t}`
 *      )
 * )
 *
 * @category instance operations
 * @since 1.0.0
 */
export const fold =
  <E, A, B>(
    loading: () => B,
    failure: (e: E) => B,
    empty: () => B,
    success: (a: A) => B
  ) => {
    return foldW<E, A, B, B, B, B>(loading, failure, empty, success)
  }
